perf-event/src/events.rs

```diff
@@ -47,22 +52,46 @@ pub enum Event {
 
     #[allow(missing_docs)]
     Cache(Cache),
+
+    #[allow(missing_docs)]
+    Breakpoint(Breakpoint),
 }
 
 impl Event {
-    pub(crate) fn r#type(&self) -> bindings::perf_type_id {
+    pub(crate) fn update_attrs(self, attr: &mut bindings::perf_event_attr) {
         match self {
-            Event::Hardware(_) => bindings::PERF_TYPE_HARDWARE,
-            Event::Software(_) => bindings::PERF_TYPE_SOFTWARE,
-            Event::Cache(_) => bindings::PERF_TYPE_HW_CACHE,
-        }
-    }
-
-    pub(crate) fn config(self) -> u64 {
-        match self {
-            Event::Hardware(hw) => hw as _,
-            Event::Software(sw) => sw as _,
-            Event::Cache(cache) => cache.as_config(),
+            Event::Hardware(hw) => {
+                attr.type_ = bindings::PERF_TYPE_HARDWARE;
+                attr.config = hw as _;
+            }
+            Event::Software(sw) => {
+                attr.type_ = bindings::PERF_TYPE_SOFTWARE;
+                attr.config = sw as _;
+            }
+            Event::Cache(cache) => {
+                attr.type_ = bindings::PERF_TYPE_HW_CACHE;
+                attr.config = cache.as_config();
+            }
+            Event::Breakpoint(bp) => {
+                attr.type_ = bindings::PERF_TYPE_BREAKPOINT;
+                // Clear config in case it was set by a previous call to update_attrs
+                attr.config = 0;
+
+                match bp {
+                    Breakpoint::Data { access, addr, len } => {
+                        attr.bp_type = access.bits();
+                        attr.__bindgen_anon_3.bp_addr = addr;
+                        attr.__bindgen_anon_4.bp_len = len;
```

Comment:
Wow, this really is the only way to set that field. That seems really fragile, although I'm not sure what else bindgen could do with a transparent union.

Reply:
Well... obviously, it should generate a setter on the enclosing struct.